File: sdk/typescript/src/modules/node.ts
import type { ChainClient } from "../infrastructure/chain-client.js";
import type { SigningService } from "../infrastructure/signing-service.js";
import type {
  NodeInfoResponse,
  NodeSearchResponse,
  NodeSummary,
  NodeStatus,
} from "../types/responses.js";
import { formatKkot } from "../utils/denom.js";

const STATUS_MAP: Record<number, NodeStatus> = {
  0: "UNSPECIFIED",
  1: "REGISTERED",
  2: "ACTIVE",
  3: "INACTIVE",
  4: "JAILED",
};

function toNodeStatus(status: number | string): NodeStatus {
  if (typeof status === "number") {
    return STATUS_MAP[status] ?? "UNSPECIFIED";
  }
  const upper = status.toUpperCase();
  if (
    upper === "REGISTERED" ||
    upper === "ACTIVE" ||
    upper === "INACTIVE" ||
    upper === "JAILED" ||
    upper === "UNSPECIFIED"
  ) {
    return upper as NodeStatus;
  }
  // Try parsing enum prefix like "NODE_STATUS_ACTIVE"
  const stripped = upper.replace("NODE_STATUS_", "");
  return (STATUS_MAP[
    Object.entries(STATUS_MAP).find(([, v]) => v === stripped)?.[0] as unknown as number
  ] ?? "UNSPECIFIED");
}

export class NodeModule {
  constructor(
    private readonly chainClient: ChainClient,
    private readonly signingService: SigningService,
  ) {}

  async getInfo(nodeId?: string): Promise<NodeInfoResponse> {
    const effectiveNodeId = nodeId ?? (await this.resolveOwnNodeId());

    const response = await this.chainClient.queryRest<{
      node: {
        id: string;
        operator: string;
        agent_address: string;
        status: number | string;
        description: string;
        website: string;
        tags: string[];
        agent_share: string;
        validator_address: string;
        registered_at: string;
      };
    }>(`/seocheon/node/v1/nodes/${effectiveNodeId}`);

    const node = response.node;

    // Fetch staking info for delegation data
    let totalDelegation = "0";
    let selfDelegation = "0";
    let commissionRate = "0";

    try {
      const validator = await this.chainClient.queryRest<{
        validator: {
          tokens: string;
          commission: {
            commission_rates: { rate: string };
          };
        };
      }>(
        `/cosmos/staking/v1beta1/validators/${node.validator_address}`,
      );
      totalDelegation = validator.validator.tokens;
      commissionRate = validator.validator.commission.commission_rates.rate;

      const selfDel = await this.chainClient.queryRest<{
        delegation_response: {
          balance: { amount: string };
        };
      }>(
        `/cosmos/staking/v1beta1/validators/${node.validator_address}/delegations/${node.operator}`,
      );
      selfDelegation = selfDel.delegation_response.balance.amount;
    } catch {
      // Validator may not exist yet for REGISTERED nodes
    }

    return {
      node_id: node.id,
      operator: node.operator,
      agent_address: node.agent_address,
      status: toNodeStatus(node.status),
      description: node.description ?? "",
      website: node.website ?? "",
      tags: node.tags ?? [],
      commission_rate: commissionRate,
      agent_share: node.agent_share,
      total_delegation: formatKkot(totalDelegation),
      self_delegation: formatKkot(selfDelegation),
      validator_address: node.validator_address,
      registered_at: parseInt(node.registered_at, 10),
    };
  }

  async search(
    tag?: string,
    status?: string,
    limit?: number,
    orderBy?: string,
  ): Promise<NodeSearchResponse> {
    const effectiveLimit = limit ?? 20;
    const effectiveOrder = orderBy ?? "delegation";

    // Query nodes
    let rawNodes: Array<{
      id: string;
      status: number | string;
      tags: string[];
      description: string;
      validator_address: string;
      registered_at: string;
    }>;

    if (tag) {
      const response = await this.chainClient.queryRest<{
        nodes: typeof rawNodes;
      }>(`/seocheon/node/v1/nodes/by-tag/${tag}`);
      rawNodes = response.nodes ?? [];
    } else {
      const response = await this.chainClient.queryRest<{
        nodes: typeof rawNodes;
      }>("/seocheon/node/v1/nodes");
      rawNodes = response.nodes ?? [];
    }

    // SDK-level status filter
    if (status) {
      const targetStatus = status.toUpperCase();
      rawNodes = rawNodes.filter(
        (n) => toNodeStatus(n.status) === targetStatus,
      );
    }

    // Fetch delegation for sorting
    const nodesWithDelegation = await Promise.all(
      rawNodes.map(async (n) => {
        let delegation = "0";
        try {
          const validator = await this.chainClient.queryRest<{
            validator: { tokens: string };
          }>(
            `/cosmos/staking/v1beta1/validators/${n.validator_address}`,
          );
          delegation = validator.validator.tokens;
        } catch {
          // Validator may not exist
        }
        return { ...n, _delegation: BigInt(delegation) };
      }),
    );

    // Sort
    if (effectiveOrder === "delegation") {
      nodesWithDelegation.sort((a, b) =>
        a._delegation > b._delegation ? -1 : a._delegation < b._delegation ? 1 : 0,
      );
    } else {
      nodesWithDelegation.sort(
        (a, b) => parseInt(b.registered_at, 10) - parseInt(a.registered_at, 10),
      );
    }

    const totalCount = nodesWithDelegation.length;
    const limited = nodesWithDelegation.slice(0, effectiveLimit);

    const nodes: NodeSummary[] = limited.map((n) => ({
      node_id: n.id,
      status: toNodeStatus(n.status),
      tags: n.tags ?? [],
      total_delegation: formatKkot(n._delegation.toString()),
      description: n.description ?? "",
    }));

    return { nodes, total_count: totalCount };
  }

  private async resolveOwnNodeId(): Promise<string> {
    const address = await this.signingService.getAddress();
    const response = await this.chainClient.queryRest<{
      node: { id: string };
    }>(`/seocheon/node/v1/nodes/by-agent/${address}`);
    return response.node.id;
  }
}

File: sdk/python/seocheon/modules/node.py
"""Node module for the Seocheon SDK."""

from __future__ import annotations

from seocheon.errors.errors import ErrNodeNotFound, ErrQueryFailed
from seocheon.infrastructure.chain_client import ChainClient
from seocheon.infrastructure.signing.service import SigningService
from seocheon.internal.tx.chain_adapter import ChainClientAdapter
from seocheon.internal.tx.messages import MsgConfirmDelegation
from seocheon.internal.tx.pipeline import PipelineConfig, default_pipeline_config, execute_tx
from seocheon.internal.tx.types import TxRequest
from seocheon.types.enums import NodeStatus
from seocheon.types.responses import (
    DelegationStatusResponse,
    NodeInfoResponse,
    NodeSearchResponse,
    NodeSummary,
    TxResultResponse,
    EventAttribute,
    TxEvent,
)
from seocheon.utils.convert import format_kkot


class NodeModule:
    """Provides node-related operations."""

    def __init__(self, client: ChainClient, signer: SigningService, chain_id: str = "seocheon-1") -> None:
        self._client = client
        self._signer = signer
        self._tx_querier = ChainClientAdapter(client)
        self._tx_config: PipelineConfig = default_pipeline_config(chain_id)

    async def get_info(self, node_id: str = "") -> NodeInfoResponse:
        """Return detailed information about a node."""
        effective_node_id = node_id or await self._resolve_own_node_id()

        path = f"/seocheon/node/v1/nodes/{effective_node_id}"
        try:
            data = await self._client.query_rest(path)
        except Exception as e:
            raise ErrNodeNotFound from e

        n = data.get("node", {})
        total_delegation = "0"
        self_delegation = "0"
        commission_rate = "0"

        val_addr = n.get("validator_address", "")
        if val_addr:
            total_delegation, commission_rate = await self._query_validator_info(val_addr)
            self_delegation = await self._query_self_delegation(n.get("operator", ""), val_addr)

        return NodeInfoResponse(
            node_id=n.get("id", ""),
            operator=n.get("operator", ""),
            agent_address=n.get("agent_address", ""),
            status=NodeStatus.from_int(n.get("status", 0)),
            description=n.get("description", ""),
            website=n.get("website", ""),
            tags=n.get("tags", []),
            commission_rate=commission_rate,
            agent_share=n.get("agent_share", "0.2"),
            total_delegation=total_delegation,
            self_delegation=self_delegation,
            validator_address=val_addr,
            registered_at=int(n.get("registered_at", 0)),
        )

    async def search(
        self, tag: str = "", status: str = "", limit: int = 20, order_by: str = "delegation",
    ) -> NodeSearchResponse:
        """Find nodes matching the given criteria."""
        if limit <= 0:
            limit = 20
        if not order_by:
            order_by = "delegation"

        path = f"/seocheon/node/v1/nodes/by-tag/{tag}" if tag else "/seocheon/node/v1/nodes"
        try:
            data = await self._client.query_rest(path)
        except Exception as e:
            raise ErrQueryFailed from e

        nodes = data.get("nodes", [])

        # Filter by status
        if status:
            nodes = [n for n in nodes if NodeStatus.from_int(n.get("status", 0)) == status]

        # Enrich with delegation info
        enriched = []
        for n in nodes:
            delegation = 0
            val_addr = n.get("validator_address", "")
            if val_addr:
                del_str, _ = await self._query_validator_info(val_addr)
                try:
                    delegation = int(del_str)
                except ValueError:
                    pass
            reg_at = int(n.get("registered_at", 0))
            enriched.append({
                "summary": NodeSummary(
                    node_id=n.get("id", ""),
                    status=NodeStatus.from_int(n.get("status", 0)),
                    tags=n.get("tags", []),
                    total_delegation=format_kkot(delegation),
                    description=n.get("description", ""),
                ),
                "delegation": delegation,
                "reg_at": reg_at,
            })

        # Sort
        if order_by == "delegation":
            enriched.sort(key=lambda x: x["delegation"], reverse=True)
        else:
            enriched.sort(key=lambda x: x["reg_at"], reverse=True)

        total_count = len(enriched)
        enriched = enriched[:limit]

        return NodeSearchResponse(
            nodes=[e["summary"] for e in enriched],
            total_count=total_count,
        )

    async def get_delegation_status(
        self, delegator_address: str, validator_address: str,
    ) -> DelegationStatusResponse:
        """Query delegation confirmation status."""
        path = f"/seocheon/node/v1/delegation-confirmation/{delegator_address}/{validator_address}"
        try:
            data = await self._client.query_rest(path)
        except Exception as e:
            raise ErrQueryFailed from e

        return DelegationStatusResponse(
            expiry_epoch=int(data.get("expiry_epoch", 0)),
            current_epoch=int(data.get("current_epoch", 0)),
            in_renewal_window=bool(data.get("in_renewal_window", False)),
            renewal_window_start=int(data.get("renewal_window_start", 0)),
        )

    async def confirm_delegation(self, validator_address: str) -> TxResultResponse:
        """Confirm delegation for a validator."""
        msg = MsgConfirmDelegation(
            delegator_address=self._signer.get_address(),
            validator_address=validator_address,
        )

        result = await execute_tx(self._tx_querier, self._signer, self._tx_config, TxRequest(message=msg))

        return TxResultResponse(
            tx_hash=result.tx_hash,
            height=result.height,
            code=result.code,
            gas_used=result.gas_used,
            gas_wanted=result.gas_wanted,
            raw_log=result.raw_log,
            events=[
                TxEvent(
                    type=e.type,
                    attributes=[EventAttribute(key=a.key, value=a.value) for a in e.attributes],
                )
                for e in result.events
            ],
        )

    async def _resolve_own_node_id(self) -> str:
        agent_addr = self._signer.get_address()
        path = f"/seocheon/node/v1/nodes/by-agent/{agent_addr}"
        try:
            data = await self._client.query_rest(path)
        except Exception as e:
            raise ErrNodeNotFound from e
        node_id = data.get("node", {}).get("id", "")
        if not node_id:
            raise ErrNodeNotFound
        return node_id

    async def _query_validator_info(self, val_addr: str) -> tuple[str, str]:
        try:
            data = await self._client.query_rest(f"/cosmos/staking/v1beta1/validators/{val_addr}")
            validator = data.get("validator", {})
            tokens = validator.get("tokens", "0")
            rate = validator.get("commission", {}).get("commission_rates", {}).get("rate", "0")
            return tokens, rate
        except Exception:
            return "0", "0"

    async def _query_self_delegation(self, operator: str, val_addr: str) -> str:
        try:
            data = await self._client.query_rest(
                f"/cosmos/staking/v1beta1/validators/{val_addr}/delegations/{operator}"
            )
            return data.get("delegation_response", {}).get("balance", {}).get("amount", "0")
        except Exception:
            return "0"
